### ISSUE TITLE
If condition value can change stack signature

### ISSUE THREAD

**Issue by iFreilicht** [uiua-lang/uiua#130]
This bug can be seen in one of the current examples for if:

```
?∘(.+) 0 2 3
```

Returns `5 5` when the condition is 0, but `2` when the condition is 1, so one less value, meaning the stack signature of `if` is only determined at runtime.

The workaround is to change `∘` to `⊙∘` in this case, but that becomes inconvenient for larger functions. Different stack signatures for the two functions should be allowed, but *only* if the difference between inputs and outputs is equal. In that case, `if` can automatically "upgrade" the function with less inputs so that the stack signatures are equal, but the lower values on the stack are left untouched.

If the difference between outputs and inputs is *different* between the functions (for example, `|3.2` and `|1.1`), it would be impossible for `if` to work correctly, and so that *should* be an error.

This will also allow `if`'s stack signature to always be determined correctly, as it will be equal to the function with the larger number of inputs.

**Comment by kaikalii** [comment]
Fixed in 0d9923f012f
These will now work:
```
f ← ?∘+
g ← ?∘(.+)
[f 0 2 3] # [5]
[f 1 2 3] # [2]
[g 0 2 3] # [5 5]
[g 1 2 3] # [2 3]
```
This will still not work:
```
h ← ?∘.
```

**Comment by iFreilicht** [comment]
Perfect, tried it out in a more complex file and works flawlessly. Thank you!
